### PR TITLE
Feature: Add csv exporter for reader study questions and answers

### LINKED ISSUE #961: Add a button to download answers / questions for reader studies in csv
The button should only be available for challenge editors.

### PR DESCRIPTION
**Pull request by MikeOverkamp-diag**
Closes #961

This will add two buttons to the reader study detail view: 

export_buttons

### CODE REVIEW

File: app/grandchallenge/reader_studies/views.py [hunk 1]
```diff
@@ -236,13 +240,67 @@ class ReadersUpdate(ReaderStudyUserGroupUpdateMixin):
...
+
+    @action(detail=True)
+    def export_answers(self, request, pk=None):
+        reader_study = self.get_object()
+        self._check_export_perms(request.user, reader_study)
+
+        fields = ["creator__username", "question__question_text", "answer"]
+        data = Answer.objects.filter(
+            question__reader_study=reader_study
+        ).values_list(*fields)
```
Comment (jmsmkn) [inline app/grandchallenge/reader_studies/views.py]: The `Images` field is missing here - this is required to know what images this answer is for. The best way to serialize this for the user here is to use the images `name` as this should be unique for the reader study.

File: app/tests/reader_studies_tests/test_api.py [hunk 2]
```diff
@@ -294,3 +298,157 @@ def test_answer_is_correct_type(client, answer_type, answer, expected):
...
+    processed = exporter._preprocess_data(data)
+    assert len(processed) == lines
+
+    # Unfortunately, we have to create an actual file here, as both tempfile
+    # and StringIO seem to cause issues with line endings
+    with open("/tmp/csv.csv", "w+", newline="") as f:
+        writer = csv.writer(f)
+        writer.writerows(processed)
+
+    with open("/tmp/csv.csv", "r", newline="") as f:
```
Comment (jmsmkn) [inline app/tests/reader_studies_tests/test_api.py]: `/tmp/` should not be used directly - better to use the `tmp_path` fixture from pytest http://doc.pytest.org/en/latest/tmpdir.html#the-tmp-path-fixture

File: app/grandchallenge/reader_studies/views.py [hunk 3]
```diff
@@ -236,13 +240,67 @@ class ReadersUpdate(ReaderStudyUserGroupUpdateMixin):
...
+            processed.append(
+                map(lambda x: re.sub(r"[\n\r\t]", " ", str(x)), entry)
+            )
+        return processed
+
+    def _create_csv_response(self, data, filename="export.csv"):
+        response = HttpResponse(content_type="text/csv")
+        response["Content-Disposition"] = f'attachment; filename="{filename}"'
+        writer = csv.writer(response, quoting=csv.QUOTE_ALL, escapechar="\\")
+        writer.writerows(self._preprocess_data(data))
```
Comment (jmsmkn) [inline app/grandchallenge/reader_studies/views.py]: Could we also add a header row?

File: app/grandchallenge/reader_studies/views.py (hunk 1 above)
Comment (MikeOverkamp-diag) [reply app/grandchallenge/reader_studies/views.py]: Added and moved all fields to the answers export

File: app/tests/reader_studies_tests/test_api.py (hunk 2 above)
Comment (MikeOverkamp-diag) [reply app/tests/reader_studies_tests/test_api.py]: Changed

File: app/grandchallenge/reader_studies/views.py (hunk 3 above)
Comment (MikeOverkamp-diag) [reply app/grandchallenge/reader_studies/views.py]: Added

File: app/grandchallenge/reader_studies/models.py [hunk 4]
```diff
@@ -455,6 +455,8 @@ class Question(UUIDModel):
     )
     order = models.PositiveSmallIntegerField(default=100)
 
+    csv_headers = ["Question text", "Answer type", "Required", "Image port"]
+
```
Comment (jmsmkn) [inline app/grandchallenge/reader_studies/models.py]: Last question: would it be better to use an OrderedDict for the csv headers (keys) and corresponding values? I'm worried that these 2 lists could get out of sync.

Comment (MikeOverkamp-diag) [reply app/grandchallenge/reader_studies/models.py]: Good point, I'll update it.

**Pull request merged**